features/support/env.rb

```diff
@@ -72,9 +44,30 @@
 #   Before('not @no-txn', 'not @selenium', 'not @culerity', 'not @celerity', 'not @javascript') do
 #     DatabaseCleaner.strategy = :transaction
 #   end
-#
 
-# Possible values are :truncation and :transaction
+require 'cucumber/rails'
+
+# Remove/comment out the lines below if your app doesn't have a database.
+# For some databases (like MongoDB and CouchDB) you may need to use :truncation instead.
+# Possible values are :truncation and :transaction or a class
 # The :transaction strategy is faster, but might give you threading problems.
 # See https://github.com/cucumber/cucumber-rails/blob/master/features/choose_javascript_database_strategy.feature
-Cucumber::Rails::Database.javascript_strategy = :truncation
+require_relative 'cucumber_database_null_javascript_strategy'
+Cucumber::Rails::Database.javascript_strategy = ::ThreeScale::CucumberDatabaseNullJavascriptStrategy
```

Comment:
I do not understand why we had to separate the concerns.

The cucumber database strategy should call `before_js` and such ... Those methods then set/unset the DatabaseCleaner strategy. Now you are separating concerns which is confusing.
I think the issue is the order of when those methods are called (as they are callbacks)

```ruby
# I think this should be enough:

Before non_transactional do
  Cucumber::Rails::Database.javascript_strategy = :truncation
end
```

But the callbacks should be called like this:

- Cucumber::Rails::Database.javascript_strategy = :truncation
- Cucumber::Rails::Database.strategy.before_js
- run your tests
- Cucumber::Rails::Database.strategy.after


My guess is that the order of the call of the callbacks is wrong

The issue is probable `before_js` (of transaction strategy) is called before setting the `javascript_strategy = :truncation` ... which makes it enter a transaction then truncate which will autocommit  as it is a DDL statement

This is what I debugged for MySQL. And I am deeply sorry I did not dig deeper for other DBs (really my fault) as I was supposing it was the same for other DB

Probably there is something either in the code or in the cucumber-* or rails-* or  oracle-* gems that creates this issue for Postgres and Oracle

Reply:
In-built hooks run after our defined hooks and mess-up everything. They only regard `@javascript` and `not @javascript` so our tag lists become irrelevant.